File: SLWH_Lua/Assets/Scripts/CSharpGame/GameLogic/Module/GamePlayer.cs


using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using System.Linq;

//
public class ItemInfo
{
    public int ItemID;
    public int ItemSubID;
    public long ItemCount;

    /// <summary>
    /// 构造物品
    /// </summary>
    /// <param name="item_id">物品类型ID</param>
    /// <param name="item_sub_id">物品子ID</param>
    /// <param name="item_count">物品数量</param>
    public ItemInfo(int item_id, int item_sub_id, long item_count)
    {
        ItemID = item_id;
        ItemSubID = item_sub_id;
        ItemCount = item_count;
    }
}

public class ItemUsingInfo
{
    public ItemInfo itemInfo;
    public int cooldown;
    public ulong endTime;
    public int seatId;

    public ItemUsingInfo(int item_id, int item_sub_id, long item_count, int cd, ulong end_time, int seat_id)
    {
        itemInfo = new ItemInfo(item_id, item_sub_id, item_count);
        cooldown = cd;
        endTime = end_time;
        seatId = seat_id;
    }
}

public class GamePlayer
{
    public int UserID;
    public string NickName;
    public EnumPlayerGenderType Gender;
    public int Head;
    public int HeadFrame;
    public int Level;
    public long LevelExp;
    public int VIPLevel;
    public long VIPLevelExp;
    public string Phone;
    public long Currency;
    public long BindCurrency;
    public long Diamond;
    public long integral;
    public List<ItemInfo> ItemList;
    public int IsChangeName;
    public int GuildId;
    //是否有未处理的申请
    public int GuildJoinListState;
    public long UnLockGun;
    public int GunId;
    public uint MonthCardExpireTime;
    public int MonthCardHasFetched;
    public int FinishFirstRecharge;
    public int ReliefFinishCount;
    public int HasUnReadMail;
    public int FetchedFirstPackage;

    public GamePlayer(CLGT.LoginAck ack)
    {
        UserID = ack.UserId;
        NickName = ack.Nickname;
        IsChangeName = ack.NicknameMdf?1:0;
        Gender = (EnumPlayerGenderType)ack.Gender;
        Head = ack.Head;
        HeadFrame = ack.HeadFrame;
        Level = ack.Level;
        LevelExp = ack.LevelExp;
        VIPLevel = ack.VipLevel;
        VIPLevelExp = ack.VipLevelExp;
        Phone = ack.Phone;
        Currency = ack.Currency;
        BindCurrency = ack.BindCurrency;
        Diamond = ack.Diamond;
        integral = ack.Integral;
        ItemList = new List<ItemInfo>();
        foreach (var item in ack.Items)
        {
            var t = new ItemInfo(item.ItemId, item.ItemSubId, item.ItemCount);
            var p = ItemList.Find(a => a.ItemID == t.ItemID && a.ItemSubID == t.ItemSubID);
            if (!ReferenceEquals(p, null))
                p.ItemCount = t.ItemCount;
            else
                ItemList.Add(t);
        }
        TimeHelper.SetServerTimestamp((ulong)ack.ServerTimestamp * 1000);



    }

    private bool _tryGetParameterFromJson<T>(JObject paramsContainer, string paramName, out T value)
    {
        bool success = true;
        value = default(T);
        try
        {
            JToken token = paramsContainer[paramName];
            if (token == null)
                throw new Exception("成员不存在");
            value = token.ToObject<T>();
        }
        catch (System.Exception ex)
        {
            UnityEngine.Debug.LogWarning($"解析LoginAck数据失败，'{paramName}':{ex.Message}");
            success = false;
        }
        return success;
    }

    public long Integral
    {
        get
        {
            return integral / SysDefines.ScoreViewPara;
        }
    }

    //设置金币
    public void SetCurrency(long currency)
    {
        if (Currency != currency)
        {
            var oldValue = Currency;
            Currency = currency;

            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_CURRENCY_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", currency},
                });
        }
    }

    //设置钻石
    public void SetDiamond(long diamond)
    {
        if (Diamond != diamond)
        {
            var oldValue = Diamond;
            Diamond = diamond;

            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_DIAMOND_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", diamond},
                });
        }
    }



    //获取背包中的道具
    public ItemInfo GetItem(int id, int subId)
    {
        var item = ItemList.Find(a => a.ItemID == id && a.ItemSubID == subId);
        return item;
    }

    //解锁最大炮倍数变化
    public void SetUnLockGun(long unLockGun, long delta)
    {
        if (UnLockGun != unLockGun)
        {
            var oldValue = UnLockGun;
            UnLockGun = unLockGun;

            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_UNLOCKGUN_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", unLockGun},
                    {"Delta", delta},
                });
        }
    }

    //切换炮台
    public void SwitchGun(int gunId)
    {
        if (GunId != gunId)
        {
            GunId = gunId;
        }
    }

    //资源强同步
    public void SyncResource(long diamond, long currency, long _integral)
    {
        if (Diamond != diamond)
            DeltaDiamond(diamond - Diamond);

        if (Currency != currency)
            DeltaCurrency(currency - Currency);

        if (integral != _integral)
            DeltaIntegral(_integral - integral);
    }

    //金币增量 0表示需要加 1表示不需要处理 默认大厅
    public void DeltaCurrency(long delta, int reason = 0)
    {
        if (0 != delta)
        {
            var oldValue = Currency;
            Currency += delta;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_CURRENCY_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", Currency},
                    {"Delta",    delta },
                    {"Reason",   reason},
                });
        }
    }

    /// <summary>
    /// 绑金增量
    /// </summary>
    /// <param name="delta">变化量</param>
    /// <param name="reason">0表示需要加 1表示不需要处理 默认大厅</param>
    public void DeltaBindCurrency(long delta,int reason = 0)
    {
        if (0!= delta)
        {
            var oldValue = BindCurrency;
            BindCurrency += delta;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_BINDCURRENCY_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", BindCurrency},
                    {"Delta",    delta },
                    {"Reason",   reason},
                });
        }
    }


    //钻石增量
    public void DeltaDiamond(long delta,int reason = 0)
    {
        if (0 != delta)
        {
            var oldValue = Diamond;
            Diamond += delta;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_DIAMOND_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", Diamond},
                    {"Delta",    delta},
                    {"Reason",   reason},
                });
        }
    }

    //积分增量
    public void DeltaIntegral(long delta)
    {
        if (0 != delta)
        {
            var oldValue = Integral;
            integral += delta;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_INTEGRAL_CHANGED,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", Integral},
                });
        }
    }

    //设置新头像
    public void SetNewHead(int headId)
    {
        if (Head != headId)
        {
            var oldValue = Head;
            Head = headId;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_HEADID,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", headId},
                });
        }
    }

    //设置新昵称
    public void SetNickName(string nickName)
    {
        if (!NickName.Equals(nickName))
        {
            var oldValue = NickName;
            NickName = nickName;
            IsChangeName = 1;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_NICKNAME,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"OldValue", oldValue},
                    {"NewValue", nickName},
                });
        }
    }

    //vip经验变化
    public void SetVipLevel(int vipLevel, long vipLevelExp)
    {
        var vipLevelChange = VIPLevel != vipLevel;
        VIPLevel = vipLevel;
        VIPLevelExp = vipLevelExp;
        MessageCenter.Instance.SendMessage(
            MsgType.CLIENT_PLAYER_VIPLEVEL,
            this,
            null,
            new Dictionary<string, object>()
            {
                {"VIPLevel", vipLevel},
                {"VIPLevelExp", vipLevelExp},
                {"VIPLevelChange", vipLevelChange},
            });
    }

    //公会ID变化
    public void SetGuildId(int id)
    {
        if (GuildId != id)
        {
            GuildId = id;
            MessageCenter.Instance.SendMessage(
                MsgType.CLIENT_PLAYER_GUILDID,
                this,
                null,
                new Dictionary<string, object>()
                {
                    {"NewValue", id},
                });
        }
    }

    //经验变化
    public void SetLevelExp(long exp)
    {
        LevelExp = exp;
        MessageCenter.Instance.SendMessage(
            MsgType.CLIENT_PLAYER_LEVELEXP,
            this,
            null,
            new Dictionary<string, object>()
            {
                {"Exp",exp}
            });
    }


}

public class TaskInfo
{
    public int MainType;            //任务主类型: 1普通任务 2成就任务
    public int Id;                  //任务Id
    public int TaskType;            //任务类型: 普通任务(1每日任务 2每周任务) 成就任务(1累计登录 2捕鱼能手 3倍率达人 4竞技高手)
    public int CompleteNum;         //完成数量
    public int NeedNum;             //需完成数量
    public string Desc;             //任务描述
    public List<ItemInfo> Rewards;  //奖励
    public bool CompleteFlag;       //是否已完成并领取奖励

    public TaskInfo(int mainType, int id, int taskType, int completeNum, int needNum, string desc, List<ItemInfo> rewards)
    {
        MainType = mainType;
        Id = id;
        TaskType = taskType;
        CompleteNum = completeNum;
        NeedNum = needNum;
        Desc = desc;
        Rewards = rewards;
        CompleteFlag = false;
    }
}

public class ActiveInfo
{
    public int DailyActiveValue;        //日活跃值
    public int WeeklyActiveValue;       //周活跃值
    public List<int> CompleteIdList;    //已领取奖励

    public ActiveInfo(int daily, int weekly, List<int> complete)
    {
        DailyActiveValue = daily;
        WeeklyActiveValue = weekly;
        CompleteIdList = complete;
    }
}

public class ShakeNumberInfo
{
    public int NumberLength;            //已摇了几天的数字
    public List<int> NumberList;        //已摇到的数字(下标0-6表示1-7天)
    public List<int> BoxStateList;      //宝箱状态,下标同上,0未领取1已领取2未解锁
    public int ActFlag;                 //当日是否已摇数字 1是0否
    public int FetchedFlag;             //本轮是否已领取 1是0否2未解锁

    public ShakeNumberInfo(int numberLength, List<int> numberList, List<int> boxStateList, int actFlag, int fetchedFlag)
    {
        NumberLength = numberLength;
        NumberList = numberList;
        BoxStateList = boxStateList;
        ActFlag = actFlag;
        FetchedFlag = fetchedFlag;
    }
}

public class WelfarePigInfo
{
    public int Welfare;
    public uint ExpireTime;
    public int IsFetched;
    public int IsBroken;
    public int BrokenCount;

    public WelfarePigInfo(int welfare, uint expireTime, int isFetched, int isBroken, int brokenCount)
    {
        Welfare = welfare;
        ExpireTime = expireTime;
        IsFetched = isFetched;
        IsBroken = isBroken;
        BrokenCount = brokenCount;
    }
}

public class InvestGunInfo
{
    public int MaxRechargeId;
    public int MaxGunValue;
    public List<int> FinishedList;

    public InvestGunInfo(int maxRechargeId, int maxGunValue, int[] finishedArray)
    {
        MaxRechargeId = maxRechargeId;
        MaxGunValue = maxGunValue;
        FinishedList = new List<int>();
        Array.ForEach(finishedArray, f => { FinishedList.Add(f); });
    }
}

public class InvestCostInfo
{
    public int IsRecharged;         //是否已完成充值 1是0否
    public long TotalCost;
    public List<int> FinishedList;

    public InvestCostInfo(int isRecharged, long totalCost, int[] finishedArray)
    {
        IsRecharged = isRecharged;
        TotalCost = totalCost;
        FinishedList = new List<int>();
        Array.ForEach(finishedArray, f => { FinishedList.Add(f); });
    }
}

File: Assets/Scripts/Lua/Module/GamePlayer.lua
local GS = GS
local GF = GF
local g_Env = g_Env or {}
local AssertUnityObjValid = AssertUnityObjValid or {}

local math, Log, tostring, typeof, debug, string, assert, coroutine, tonumber
    = math, Log, tostring, typeof, debug, string, assert, coroutine, tonumber

local luaItemInfo = require 'UI.Item.LuaItemInfo'
local ItemT = require 'Table.Item'

local Class = class(nil, {
    UserID = 0,
    NickName = "",
    Gender = 0, -- 0保密 1男 2女
    Head = 0,
    HeadFrame = 0,
    Level = 0,
    LevelExp = 0,
    VIPLevel = 0,
    VIPLevelExp = 0,
    Phone = "",
    Currency = 0,
    BindCurrency = 0,
    Diamond = 0,
    integral = 0,
    ItemList = {},
    IsChangeName = 0, -- 0未改过名字 1改过
    GuildId = 0,
    GuildJoinListState = false,
    UnLockGun = 0, -- 最大炮倍
    GunId = 0,
    MonthCardExpireTime = 0, -- 月卡过期时间（时间戳，单位秒）
    MonthCardHasFetched = false, -- 月卡是否领取
    FinishFirstRecharge = false, -- 首冲购买
    ReliefFinishCount = 0, -- 救济金领取次数
    HasUnReadMail = false, -- 有没有未读邮件
    HasUnReadFeedback = false, -- 有没有未读的反馈回复
    HasOnlineService = false, -- 有没有在线客服回复消息
    HasNewQA = false, -- 常见问题配置表有没有更新
    FetchedFirstPackage = false,
    MultipleHit = 0, -- 倍击倍数
    BankPasswordFlag = false, -- 银行密码是否重置过
    LaserGunEnergy = 0, -- 激光炮能量
    RunningGameServer = "", -- 当前进入的游戏服务器名称
    RankRewardGold = false, -- 排行榜金币奖励是否领取
    RankRewardWarhead = false, -- 排行榜弹头奖励是否领取
    MagicValue = 0, -- 魔力值
    MagicItemId = 0,
    MagicFactor = 0,
    BankCurrency = 0, -- 银行资产
    customHeadData = {},
    HeadUrl = '',
    Token = '',
})

function Class:SetData(ack)
    self:__init(ack)
    Log("GamePlayer SetData NickName = ", self.NickName, self.Level)
end

function Class:__init(ack)
    if ack == nil then
        return
    end
    self.UserID = ack.user_id
    self.NickName = ack.nickname
    self.IsChangeName = ack.nickname_mdf and 1 or 0
    self.Gender = ack.gender
    self.Head = ack.head
    self.HeadFrame = ack.head_frame
    self.Level = ack.level
    self.LevelExp = ack.level_exp
    self.VIPLevel = ack.vip_level
    self.VIPLevelExp = ack.vip_level_exp
    self.Phone = ack.phone
    self.Currency = ack.currency
    self.BankCurrency = ack.bank_currency
    self.BindCurrency = ack.bind_currency
    self.Diamond = ack.diamond
    self.integral = ack.integral
    self.HeadUrl = ack.head_url
    for i = 1, #(ack.items) do
        local item = ack.items[i]
        local itemInfo = luaItemInfo(item.item_id, item.item_sub_id, item.item_count)
        table.insert(self.ItemList, itemInfo)
    end
    GS.TimeHelper.SetServerTimestamp(ack.server_timestamp * 1000)
    self.GuildId = ack.guild_id
    self.GuildJoinListState = ack.guild_join_list_state
    self.MonthCardExpireTime = ack.month_card_expire_time
    self.MonthCardHasFetched = ack.month_card_has_fetched
    self.FinishFirstRecharge = ack.finish_first_recharge
    self.ReliefFinishCount = ack.relief_finish_count
    self.HasUnReadMail = ack.has_unread_mail
    self.FetchedFirstPackage = ack.fetched_first_package
    local max_gun_value = ack.max_gun_value
    -- local fistGun = GF.table.FindBy(FishGunForgeT, function(g)
    --     return g.Id == 1
    -- end)
    self.UnLockGun = max_gun_value --math.max(max_gun_value, fistGun.GunValue)
    self.BankPasswordFlag = ack.bank_password_flag
    self.RankRewardGold = ack.rank_reward_gold
    self.RankRewardWarhead = ack.rank_reward_warhead
    self.MagicValue = ack.magic_value
    self.MagicItemId = ack.magic_item_id
    self.MagicItemSubId = ack.magic_item_sub_id
    self.MagicFactor = ack.magic_factor
    -- Log("打印出luagamePlayer:"..GF.table.Log(self))
end
-- 设置金币
function Class:SetCurrency(currency)
    if self.Currency ~= currency then
        local _oldValue = self.Currency
        self.Currency = currency
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.Currency
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_CURRENCY_CHANGED", msg)
    end
end
-- 设置银行资产
function Class:SetBankCurrency(bankCurrency)
    if self.BankCurrency ~= bankCurrency then
        local _oldValue = self.BankCurrency
        self.BankCurrency = bankCurrency
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.BankCurrency
        }
        g_Env.messageCenter:SendMessage("CLIENT_BANK_CURRENCY_CHANGED", msg)
    end
end

-- 设置钻石
function Class:SetDiamond(diamond)
    if self.Diamond ~= diamond then
        local _oldValue = self.Diamond
        self.Diamond = diamond
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.Diamond
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_DIAMOND_CHANGED", msg)
    end
end
-- 道具变化
function Class:SetItem(items)
    if GF.table.nums(items) > 0 then
        for i = 1, #items do
            local t = luaItemInfo(items[i].item_id, items[i].item_sub_id, items[i].item_count)
            local item = GF.table.FindBy(self.ItemList, function(a)
                return a.ItemID == t.ItemID and a.ItemSubID == t.ItemSubID
            end)
            if item ~= nil then
                item.ItemCount = t.ItemCount
            else
                table.insert(self.ItemList, t)
            end
            local _content = ItemT[items[i].item_id][items[i].item_sub_id]

            g_Env.messageCenter:SendMessage("CLIENT_PLAYER_ITEM_CHANGED", {
                sender = self,
                content = _content
            })
        end
    end
end
-- 获取背包道具
function Class:GetItem(id, subId)
    if #(self.ItemList) == 0 then
        return nil
    else
        local item = GF.table.FindBy(self.ItemList, function(a)
            return a.ItemID == id and a.ItemSubID == subId
        end)
        return item
    end
end
-- 解锁最大炮倍变化
function Class:SetUnLockGun(unLockGun, delta)
    if self.UnLockGun ~= unLockGun then
        local _oldValue = self.UnLockGun
        self.UnLockGun = unLockGun
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = unLockGun,
            delta = delta
        }
        Log("玩家炮台倍数:" .. tostring(self.UnLockGun))
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_UNLOCKGUN_CHANGED", msg)
    end
end

-- 切换炮台
function Class:SwitchGun(gunId)
    if self.GunId ~= gunId then
        self.GunId = gunId
    end
end
-- 强同步
function Class:SyncResource(diamond, currency, integral)
    if self.Diamond ~= diamond then
        self.Diamond = diamond
    end
    if self.Currency ~= currency then
        self.Currency = currency
    end
    if self.integral ~= integral then
        self:DeltaIntegral(integral - self.integral)
    end
end

-- 钻石变化
function Class:DeltaDiamond(delta, _reason)
    if delta ~= 0 then
        if _reason == nil then
            _reason = 0
        end
        local _oldValue = self.Diamond
        self.Diamond = self.Diamond + delta
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.Diamond,
            delta = delta,
            reason = _reason
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_DIAMOND_CHANGED", msg)
    end
end
-- 金币变化
function Class:DeltaCurrency(delta, _reason)
    if delta ~= 0 then
        if _reason == nil then
            _reason = 0
        end
        local _oldValue = self.Currency
        self.Currency = self.Currency + delta
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.Currency,
            delta = delta,
            reason = _reason
        }

        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_CURRENCY_CHANGED", msg)

        --给小游戏发送(兼容措施)
        if GS.MessageCenter and GS.MessageCenter.instance then
            local ppt = {
                Reason = _reason,
                Delta = delta,
                NewValue = self.Currency,
                OldValue = _oldValue,
            }
            msg.get_Item = function (_,param)
                if param=="Params" then
                    return {
                        get_Item = function (_,p)
                            if p=='0' then
                                return {
                                    get_Item = function (_,pp)
                                        return ppt[pp]
                                    end
                                }
                            end
                        end
                    }
                end
                
            end
            GS.MessageCenter.instance:SendMessage("CLIENT_PLAYER_CURRENCY_CHANGED", msg)
        end
    end
end

function Class:DeltaBankCurrency(delta, _reason)
    if delta ~= 0 then
        local _oldValue = self.BankCurrency
        self.BankCurrency = self.BankCurrency + delta
        -- Log("玩家增加的银行金币数:" .. tostring(delta))
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.BankCurrency,
            delta = delta
        }
        g_Env.messageCenter:SendMessage("CLIENT_BANK_CURRENCY_CHANGED", msg)
    end
end
-- 绑定金币变化
function Class:DeltaBindCurrency(delta, _reason)
    if delta ~= 0 then
        if _reason == nil then
            _reason = 0
        end
        local _oldValue = self.BindCurrency
        self.BindCurrency = self.BindCurrency + delta
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.BindCurrency,
            delta = delta,
            reason = _reason
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_BINDCURRENCY_CHANGED", msg)
    end
end
-- 积分变化
function Class:DeltaIntegral(delta, _reason)
    if delta ~= 0 then
        if _reason == nil then
            _reason = 0
        end
        local _oldValue = self.integral
        self.integral = self.integral + delta
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.integral,
            delta = delta,
            reason = _reason
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_INTEGRAL_CHANGED", msg)
    end
end

-- 设置头像
function Class:SetNewHead(headId)
    if headId ~= self.Head then
        local _oldValue = self.Head
        self.Head = headId
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = headId
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_HEADID", msg)
        g_Env.accountListManager:UpdateAccountHead(headId)
    end
end
-- 设置头像框
function Class:SetNewFrame(frameId)
    if frameId ~= self.HeadFrame then
        local _oldValue = self.HeadFrame
        self.HeadFrame = frameId
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = frameId
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_FRAMEID", msg)
        g_Env.accountListManager:UpdateAccountFrame(frameId)
    end
end
-- 设置昵称
function Class:SetNickName(nickName)
    if nickName ~= self.NickName then
        local _oldValue = self.NickName
        self.NickName = nickName
        self.IsChangeName = 1
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = nickName
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_NICKNAME", msg)
        g_Env.accountListManager:UpdateAccountNickname(nickName)
    end
end
-- vip经验变化
function Class:SetVipLevel(vipLevel, vipLevelExp)
    local vipLevelChange = (self.VIPLevel ~= vipLevel)
    self.VIPLevel = vipLevel
    self.VIPLevelExp = vipLevelExp
    local msg = {
        sender = self,
        VIPLevel = vipLevel,
        VIPLevelExp = vipLevelExp,
        VIPLevelChange = vipLevelChange
    }
    g_Env.messageCenter:SendMessage("CLIENT_PLAYER_VIPLEVEL", msg)
end
-- 经验变化
function Class:SetLevelExp(exp)
    self.LevelExp = exp
    local msg = {
        sender = self,
        Exp = exp
    }
    g_Env.messageCenter:SendMessage("CLIENT_PLAYER_LEVELEXP", msg)
end
-- 等级变化
function Class:SetLevel(level, items)
    self.Level = level
    local msg = {
        sender = self,
        Level = level,
        Items = items
    }
    g_Env.messageCenter:SendMessage("CLIENT_PLAYER_LEVEL", msg)
end

-- 设置倍击倍数
function Class:SetMultipleHitRate(rate)
    self.MultipleHit = rate
    g_Env.messageCenter:SendMessage("ROOM_MUILTIPLE_HIT_CHANGE", {
        sender = self
    })
end

-- 设置激光炮能量
function Class:SetLaserEnergy(value)
    self.LaserGunEnergy = value
    g_Env.messageCenter:SendMessage("ROOM_PLAYER_LASER_ENERGY_CHANGE", {
        sender = self
    })
end

-- 能量值变化
function Class:DeltaLaserEnergy(delta)
    if delta ~= 0 then
        self.LaserGunEnergy = self.LaserGunEnergy + delta
        if self.LaserGunEnergy < 0 then
            self.LaserGunEnergy = 0
        end
        g_Env.messageCenter:SendMessage("ROOM_PLAYER_LASER_ENERGY_CHANGE", {
            sender = self
        })
    end
end
-- 设置魔力值
function Class:SetMagicValue(value)
    if self.MagicValue ~= value then
        local _oldValue = self.MagicValue
        self.MagicValue = value
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = value
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_MAGIC_CHANGED", msg)
    end
end
-- 能量值变化
function Class:DeltaMagicValue(delta, _reason)
    if delta ~= 0 then
        if _reason == nil then
            _reason = 0
        end
        local _oldValue = self.MagicValue
        self.MagicValue = self.MagicValue + delta
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = self.MagicValue,
            delta = delta,
            reason = _reason
        }
        g_Env.messageCenter:SendMessage("CLIENT_PLAYER_MAGIC_CHANGED", msg)
    end
end

-- 月卡时间变化
function Class:SetMonthCardExpireTime(value)
    if self.MonthCardExpireTime ~= value then
        local _oldValue = self.MonthCardExpireTime
        self.MonthCardExpireTime = value
        local msg = {
            sender = self,
            oldValue = _oldValue,
            newValue = value
        }
        Log("发送月卡时间变化消息: ",self.MonthCardExpireTime, value)
        g_Env.messageCenter:SendMessage("CLIENT_MONTHCARD_REFRESH", msg)
    end
end

function Class:Release()
    self.UserID = 0
    self.NickName = ""
    self.IsChangeName = 0
    self.Gender = 0
    self.Head = 0
    self.HeadFrame = 0
    self.Level = 0
    self.LevelExp = 0
    self.VIPLevel = 0
    self.VIPLevelExp = 0
    self.Phone = ""
    self.Currency = 0
    self.BankCurrency = 0
    self.BindCurrency = 0
    self.Diamond = 0
    self.integral = 0
    GS.TimeHelper.SetServerTimestamp(0)
    self.GuildId = 0
    self.GuildJoinListState = false
    self.MonthCardExpireTime = 0
    self.MonthCardHasFetched = false
    self.FinishFirstRecharge = false
    self.ReliefFinishCount = 0
    self.HasUnReadMail = false
    self.FetchedFirstPackage = false
    self.UnLockGun = 0
    self.BankPasswordFlag = false
    self.RankRewardGold = false
    self.RankRewardWarhead = false
    self.MagicValue = 0
    self.MagicItemId = 0
    self.MagicItemSubId = 0
    self.MagicFactor = 0
end

return Class
